third_party/txt/src/txt/font_features.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LIB_TXT_SRC_FONT_FEATURES_H_
+#define LIB_TXT_SRC_FONT_FEATURES_H_
+
+#include <map>
+#include <string>
+#include <vector>
+
+namespace txt {
+
+// Feature tags that can be applied in a text style to control how a font
+// selects glyphs.
+class FontFeatures {
+ public:
+  void SetFeature(std::string tag, int value);
+
+  std::string GetFeatureSettings() const;
+
+ private:
+  class Feature {
+   public:
+    Feature();
+    Feature(std::string tag, int value);
+
+    std::string tag() const { return tag_; }
+    int value() const { return value_; }
+
+   private:
+    std::string tag_;
+    int value_;
+  };
+
+  enum FeatureType {
+    kUnknownType,
+    kNumberSpacingType,
```

Comment:
Is NumberSpacingType the only feature we currently support? Is this meant to be expanded as we support more?

Reply:
Some sets of font features are mutually exclusive.  This is intended as a way to assign features to groups so that a child style's feature will override the parent style's feature if they are in the same group.

In the current `FontFeature` API docs, we only claim to support the overriding behavior for proportional versus tabular numbers.  All other features will be given to HarfBuzz, which will apply its own logic to resolve any conflicting features.

In the future, we may define more `FeatureType`s to ensure that the engine controls the overriding behavior for other feature tags.

Reply:
why do we need this here? I would prefer it if just giving these in the right order automatically solved this for us.

Reply:
(because it would mean not having to have an opinion or knowledge about these things... I would hate for us to get into the business of having to push updates to handle new font features...)

Reply:
HarfBuzz has its own logic for resolving conflicting font features, and it does not match the override behavior specified in the `FontFeature` Dart API.  If we want that override behavior, then it will need to be explicitly implemented in the engine.

Reply:
What's HarfBuzz's behaviour? Does it match what's described on the Microsoft site?
I thought all this was supposed to be handed to the font for processing. How much does HarfBuzz do with it?

At the end of the day if HarfBuzz does something reasonable then I'm fine with just deferring to HarfBuzz (and documenting that instead of what I wrote).

Reply:
AFAICT HarfBuzz/OpenType does not promise any particular behavior if you specify conflicting font features.

I tried passing both proportional figures and tabular figures feature tags to HarfBuzz (using Roboto and GoogleSans fonts on Android).  The font rendered with tabular figures regardless of the order in which the tags were passed to HarfBuzz.
However, I don't know if that preference is guaranteed to be consistent across fonts and across versions of HarfBuzz.

Perhaps it would be simpler to specify separate attributes for mutually exclusive features instead of placing all features in one list?  The API could be something like what CSS does (https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-numeric)

Reply:
ping me on Monday, we'll try to figure this out.